website/docs/r/instance.html.markdown

```diff
@@ -73,9 +73,11 @@ The following arguments are supported:
 instance. Amazon defaults this to `stop` for EBS-backed instances and
 `terminate` for instance-store instances. Cannot be set on instance-store
 instances. See [Shutdown Behavior](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/terminating-instances.html#Using_ChangingInstanceInitiatedShutdownBehavior) for more information.
-* `instance_type` - (Required) The type of instance to start. Updates to this field will trigger a stop/start of the EC2 instance.
+* `instance_type` - (Optional) The type of instance to start. Can be set here or
+  in Launch Template. If type is not set here or in provided `launch_template` instance type will
+  default to `m1.small`. Updates to this field will trigger a stop/start of the EC2
```

Comment:
This seems suboptimal - is there an option for "required somewhere" where it won't get a default?

Reply:
Currently, there is no way in terraform to require one of two fields to be set during plan time. So we either can faile during apply time if both settings are empty or rely on API defaults. What do you think? 

I'm not sure where I've previously gotten info about `m1.small` being  default for launch templates, might not be true anymore. I seems to be a default for Cloud formation https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html

Reply:
@gordonbondon the Terraform Plugin SDK recently added [`ExactlyOneOf` to the `Schema` type](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/schema#Schema):

>     // ExactlyOneOf is a set of schema keys that, when set, only one of the
>     // keys in that list can be specified. It will error if none are
>     // specified as well.

Its usage is a little awkward since its per-attribute and not at the resource level, but adding it to both schema attributes (e.g. `ExactlyOneOf: []string{"instance_type", "launch_template"}`) should automatically trigger a plan-time error.

Reply:
@bflad great! I'll add it. I also needed this for `ami_id`. 

Though there's still a case where `launch_template` is provided, but no instance type is defined in this launch template - then it will also use defaults from RunInstances api https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html which is `m1.small` right now.

Reply:
@bflad @007 I've updated resource to use `AtLeastOneOf` for `ami` and `instance_type`. AWS API allows overriding both of this params when launching instance from template ([docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-launch-templates.html#launch-instance-from-launch-template)), so they can be set together with launch template specification and I wanted to preserve this behaviour. 

Resource will fail plan time if `ami` is not set and it's missing in template. For `instance_type` it will be set to API default ([doc](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html)). I guess I can add log warning mentioning this.